### ISSUE TITLE
Versioning Policy Update

### ISSUE THREAD

**Issue by speediedan** [Lightning-AI/pytorch-lightning#21073]
### Outline & Motivation

Firstly, thanks to Lightning team for your continued hard work and invaluable contributions to the open-source ecosystem! 

As a follow-up to [this issue](https://github.com/Lightning-AI/pytorch-lightning/issues/20670),  I’ve noticed that recent Lightning minor releases are no longer synchronized with PyTorch’s minor releases as they have been in the past:

| PyTorch Version | Release Date   | Release Artifact                                                                | Lightning Version | Release Date      | Release Artifact                                                                                     | Lag (weeks: Lightning - PyTorch) |
|:---------------:|:--------------|:--------------------------------------------------------------------------------|:-----------------:|:------------------|:-----------------------------------------------------------------------------------------------------|:---------------------------------:|
| 2.0             | Mar 15, 2023  | [PyTorch 2.0.0](https://github.com/pytorch/pytorch/releases/tag/v2.0.0)         | 2.0.0             | Mar 15, 2023      | [Lightning 2.0.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.0.0)              | 0                                |
| 2.1             | Oct 4, 2023   | [PyTorch 2.1.0](https://github.com/pytorch/pytorch/releases/tag/v2.1.0)         | 2.1.0             | Oct 12, 2023      | [Lightning 2.1.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.1.0)              | 1                                |
| 2.2             | Jan 30, 2024  | [PyTorch 2.2.0](https://github.com/pytorch/pytorch/releases/tag/v2.2.0)         | 2.2.0             | Feb 7, 2024       | [Lightning 2.2.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.2.0)              | 1                                |
| 2.3             | Apr 24, 2024  | [PyTorch 2.3.0](https://github.com/pytorch/pytorch/releases/tag/v2.3.0)         | 2.3.0             | Jun 13, 2024      | [Lightning 2.3.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.3.0)              | 7 (50 days)                      |
| 2.4             | Jul 24, 2024  | [PyTorch 2.4.0](https://github.com/pytorch/pytorch/releases/tag/v2.4.0)         | 2.4.0             | Aug 7, 2024       | [Lightning 2.4.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.4.0)              | 2 (14 days)                      |
| 2.5             | Oct 17, 2024  | [PyTorch 2.5.0](https://github.com/pytorch/pytorch/releases/tag/v2.5.0)         | 2.5.0             | Dec 20, 2024      | [Lightning 2.5.0](https://github.com/Lightning-AI/pytorch-lightning/releases/tag/2.5.0)              | 9 (64 days)                      |
| 2.6             | Jan 29, 2025  | [PyTorch 2.6.0](https://github.com/pytorch/pytorch/releases/tag/v2.6.0)         | —                 | —                 | —                                                                                                    | 28 (197 days)                    |
| 2.7             | Apr 23, 2025  | [PyTorch 2.7.0](https://github.com/pytorch/pytorch/releases/tag/v2.7.0)         | —                 | —                 | —                                                                                                    | 16 (113 days)                    |
| 2.8             | Aug 6, 2025   | [PyTorch 2.8.0](https://github.com/pytorch/pytorch/releases/tag/v2.8.0)         | —                 | —                 | —                                                                                                    | 1 (8 days)                       |

### Pitch

I think both users and maintainers of downstream packages depending on Lightning (I'm the primary maintainer of a downstream Lightning package [Fine-Tuning Scheduler](https://github.com/speediedan/finetuning-scheduler)) would benefit if this apparent versioning policy/cadence change were made explicit in the [versioning documentation](https://github.com/Lightning-AI/pytorch-lightning/blob/a9deadaf418ae85201a5e4b91642a50c82bfd328/docs/source-pytorch/versioning.rst). 

Can we update the versioning policy to clarify the current minor release policy/cadence for Lightning? E.g., _'minor releases are typically driven by key upstream PyTorch deprecations etc...'_ This would help the community better plan dependencies and upgrades. 

### Additional context

As a concrete example, Fine-Tuning Scheduler has a number of deprecations and enhancements aligned with various PyTorch minor versions that have been awaiting a Lightning minor release. If the release cadence of the last year is expected to be the new approach moving forward, it would be good to know so FTS can proceed with changing its release approach from being Lightning release synchronized to a PyTorch release alignment instead.
Thanks again!

@Borda @lantiga @iwr-redmond 

cc @lantiga @borda @justusschock

**Comment by Borda** [comment]
Hello, thank you for your suggestion. Could you please elaborate on what issue it shall solve?
so Lightning has a policy to bump the minimal Torch version to keep about 5 versions, but there is no restriction on the latest Torch version; it is just at the time of each release of PL we cannot predict Torch's future changes so we react to them and fix them in our next patch release... It is not like Torch and TorchVision, where there is only one specific version supported.

**Comment by speediedan** [comment]
Sure thanks for weighing in @Borda. I definitely recognize the motivation for the current approach and appreciate that it doesn't prevent the latest PyTorch from being used with Lightning.

When Lightning previously released minor releases synchronized with PyTorch it allowed for faster evolution of the API, reducing cost of maintenance/CI along and facilitating adoption of research and best practices. 
I recognize there's broadly a trade-off between API stability and maintenance/CI -- slower API evolution requiring more supported previous PyTorch versions with fewer deprecation and breaking change opportunities.

To help downstream packages plan deprecations and breaking changes, I was hoping we could just add an explicit clarification/guidance regarding the number of supported PyTorch minor versions.

Concretely, there are a LOT of composable distributed API changes with torch across PyTorch 2.4 and 2.5 so FTS has to keep substantial additional code to handle cases across those versions that can be removed once 2.5 is the minimum.

Right now, the Lightning dev branch supports [PyTorch >= 2.1](https://github.com/Lightning-AI/pytorch-lightning/blob/8d1a73475e0f11ed3b3c5907da54fc993717d492/requirements/pytorch/base.txt#L4). Given we're working against PyTorch 2.9 nightlies, that's 9 versions of PyTorch across almost 2 years that need to be supported (or 8 if the minimum is bumped to 2.2 to support 5 versions from 2.5). 

I don't have a strong opinion when it comes to how many previous versions should be supported but it would be helpful for planning purposes to have a stake in the ground regarding the number (even if it's best effort). For instance, if we say we're going to support the latest 5 pytorch minor releases, that would mean when PyTorch 2.9 is released in October, Lightning would release with a minimum supported version >= 2.5 (so `main` would have that minimum or at least `minimum-1` right now). Of course, since we probably wouldn't want to bump the minimum pytorch in a patch release, that would mean more frequent minor releases.

Anyway, downstream packages can always support a smaller range of PyTorch versions than Lightning does but as Lightning continues to mature and a become an increasingly crucial part of the ML research ecosystem, I think it'd be helpful to provide the community more explicit guidance on the expected PyTorch support (e.g. most recent 5 versions of PyTorch) and possibly some approximate release cadence guidance similar to [PyTorch](https://github.com/pytorch/pytorch/blob/30d2f98daa956843a8f7b43e809da57910540306/RELEASE.md?plain=1#L65-L81) for the same reasons PyTorch does, to encourage and facilitate downstream package integration and release planning.

Thanks again for all your fantastic contributions! 🚀

**Comment by iwr-redmond** [comment]
Thanks for tagging me @speediedan.

My main concern in #20670 was to ensure that the documentation remained in-sync with the releases. Your suggestion about adding in some general language in addition to the versioning table would help with this IMHO.

**Comment by speediedan** [comment]
> My main concern in [#20670](https://github.com/Lightning-AI/pytorch-lightning/issues/20670) was to ensure that the documentation remained in-sync with the releases. Your suggestion about adding in some general language in addition to the versioning table would help with this IMHO.

Thanks for weighing in and submitting/addressing #20670 @iwr-redmond! I agree 💯. 

I think establishing an official target number of supported PyTorch minor releases would be helpful, irrespective of what number the core Lightning team decides is preferred. We could always frame the commitment as a target to continue to grant the project some latitude.

**Comment by Borda** [comment]
> When Lightning previously released minor releases synchronized with PyTorch it allowed for faster evolution of the API, reducing cost of maintenance/CI along and facilitating adoption of research and best practices.
> I recognize there's broadly a trade-off between API stability and maintenance/CI -- slower API evolution requiring more supported previous PyTorch versions with fewer deprecation and breaking change opportunities.

That is right, but since Lightning 2.0 guarantees full back compatibility to there can be some small additions or changes, but no breaking changes?

> Concretely, there are a LOT of composable distributed API changes with torch across PyTorch 2.4 and 2.5 so FTS has to keep substantial additional code to handle cases across those versions that can be removed once 2.5 is the minimum.

So far, we are testing seems all works as expected, but good call, and shall add more GPU tests for particular versions, at least for the merge to master

**Comment by speediedan** [comment]
Thanks Jirka!

> That is right, but since Lightning 2.0 guarantees full back compatibility to there can be some small additions or changes, but no breaking changes?

This is a bit of separate issue but IIUC, the API evolution policy still allows intra-major version breaking changes with the [appropriate deprecations](https://github.com/Lightning-AI/pytorch-lightning/blob/a9deadaf418ae85201a5e4b91642a50c82bfd328/docs/source-pytorch/versioning.rst?plain=1#L60C4-L60C144):
>  The breaking change is done in version ``MAJOR.(MINOR+2).0`` (e.g. ``1.7.0``), or ``(MAJOR+1).0.0`` (e.g. ``2.0.0``), whichever comes first.

Anyway, to resolve this issue, could I open a tiny docs PR with the minimal clarification below added just above the compatibility matrix?
"Though not a strict policy, Lightning strives to officially support the latest <n> PyTorch minor releases."
(it seems `n` might be 5)
Thanks!

**Comment by Borda** [comment]
since 2.0.0 any breaking change will be `(MAJOR+1).0.0` unless it fixing critical bug

**Comment by speediedan** [comment]
> since 2.0.0 any breaking change will be `(MAJOR+1).0.0` unless it fixing critical bug

Thanks for clarifying this. I've opened [a PR](https://github.com/Lightning-AI/pytorch-lightning/pull/21107) updating the versioning documentation accordingly. I've also bumped the `2.5.x` compatibility range to be current and added a little language making the intended compatibility range more explicit (while retaining flexibility). Thanks again for taking the time to discuss this!
